examples/webgl_lights_spotlight.html

```diff
@@ -64,12 +66,29 @@
 				const ambient = new THREE.AmbientLight( 0xffffff, 0.1 );
 				scene.add( ambient );
 
-				spotLight = new THREE.SpotLight( 0xffffff, 1 );
-				spotLight.position.set( 15, 40, 35 );
-				spotLight.angle = Math.PI / 4;
+				const iesProfilesTexture = new IESTexture( [
+
+					'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/007cfb11e343e2f42e3b476be4ab684e.ies',
+
+					'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/06b4cfdc8805709e767b5e2e904be8ad.ies',
+
+					//'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/02a7562c650498ebb301153dbbf59207.ies'
+
+				] );
+
+				// iesProfilesTexture.add( 'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/007cfb11e343e2f42e3b476be4ab684e.ies' );
+				// iesProfilesTexture.add( 'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/06b4cfdc8805709e767b5e2e904be8ad.ies' );
+				iesProfilesTexture.add( 'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/02a7562c650498ebb301153dbbf59207.ies' );
+
+				scene.iesProfiles = iesProfilesTexture;
```

Comment:
Thanks very much for that. 

Apologies, I had lost track of our earlier discussion, you also had asked @Mugen87 and @mrdoob in https://github.com/mrdoob/three.js/pull/24409#discussion_r933878109 for some feedback.

I think the constant index via `#define`s is a great workaround to avoid compatibility issues.

The proposal of

```js
spotLight.iesProfileMap = await new IESLoader().loadAsync( './path/to/profile.ies' );
```

seems good, much simpler.

Under the hood we can map IES profile paths to the profile index, which for 1D IES profiles is just an index into the horizontal or vertical line of the profiles texture. This also double as a caching mechanism to avoid ever loading and parsing IES profiles multiple times.

I'll do another pass over the code to implement the above and await feedback on the [spotlight logic for handling maps and shadows](https://github.com/mrdoob/three.js/blob/4c74ddc15b6a9eae758c0bc9c55c59fcf71d6464/src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js#L98-L110) issue.

Reply:
It does seem that using a `#define` index to access a constant integer array works, though (but only in WebGL2...):

```glsl
uniform sampler2D textures[ 3 ];
const int indices[ 3 ] = int[]( 2, 1, 0 );

void main() {

  gl_FragColor = texture( textures[ indices[ 0 ] ], vec2( 0.5 ) );
    
}
```

https://jsfiddle.net/694jx3ny/1/

This still wouldn't let you only conditionally include the IES code per spotlight, though, since you can't access the constant array in a `#define` condition to check if it's - 1.

Reply:
How about this?

```glsl

  #define ACCUM0 gl_FragColor += texture( textures[ indices0 ], vec2( 0.5 ) );
  #define ACCUM1 gl_FragColor += texture( textures[ indices1 ], vec2( 0.5 ) );
  #define ACCUM2 gl_FragColor += texture( textures[ indices2 ], vec2( 0.5 ) );
  
  #define LOOP0 ACCUM2  
  #define LOOP1 ACCUM1 LOOP0
  #define LOOP2 ACCUM0 LOOP1
  
  uniform sampler2D textures[3];
		
  #define indices0 2  
  #define indices1 1
  #define indices2 0

  void main() {
        
    gl_FragColor = vec4(0);
    LOOP2;
    
  }

```

https://jsfiddle.net/uhpovm9L/36/

Reply:
This is a bit of a hack but is more along the lines of what I was thinking (https://jsfiddle.net/873rkve1/2/):

The token `UNROLLED_LOOP_INDEX` is indiscriminately replaced with the loop index when unrolling a loop so using the token in a `#define` identifier results in a new indexed identifier for each loop. For example in this case `TEST_ARR_UNROLLED_LOOP_INDEX` is unpacked into `TEST_ARR_0`, `TEST_ARR_1`, and `TEST_ARR_2` for each loop iteration respectively. This effectively lets you access array values in preprocessor macros.

I'm not sure what the project will think of this, though, but it should be significantly cleaner and more maintainable than the sorted list approach being used now imo.

```js
defines: {
  TEST_ARR_0: 2,
  TEST_ARR_1: - 1,
  TEST_ARR_2: 0,
},
  
fragmentShader: /* glsl */`
  
  uniform sampler2D textures[ 3 ];
  
  void main() {
  
	  gl_FragColor = vec4( 0.0 );
	  gl_FragColor.a = 1.0;
  
	  #pragma unroll_loop_start
	  for ( int i = 0; i < 3; i ++ ) {
  
		  #if TEST_ARR_UNROLLED_LOOP_INDEX != - 1

		  gl_FragColor += texture( textures[ TEST_ARR_UNROLLED_LOOP_INDEX ], vec2( 0.5 ) );

		  #endif
  
	  }
	  #pragma unroll_loop_end
	  
  }
`
```

unrolls into:

```glsl
void main() {

    gl_FragColor = vec4( 0.0 );
    gl_FragColor.a = 1.0;
    
    #if TEST_ARR_0 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_0 ], vec2( 0.5 ) );
    
    #endif
    
    #if TEST_ARR_1 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_1 ], vec2( 0.5 ) );
    
    #endif
    
    #if TEST_ARR_2 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_2 ], vec2( 0.5 ) );
    
    #endif

}
```

Reply:
This is much better, thanks.